lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp

```diff
@@ -6754,12 +6754,12 @@ size_t TypeSystemClang::GetIndexOfChildMemberWithName(
           llvm::StringRef field_name = field->getName();
           if (field_name.empty()) {
             CompilerType field_type = GetType(field->getType());
+            std::vector<uint32_t> save_indices = child_indexes;
```

Comment:
Hmm nevermind. IIUC the reason it's clearing the indices is that it is trying to backtrack if we failed to find the member after searching through a certain path.

Reply:
I *could* do this; before I go ahead, I would like confirmation that this is what folks really want me to do?

Reply:
Actually, I take that back.  I've been playing around with this, and *not* clearing the vector is not a viable solution, because the code the clears the vector (when an error condition is encountered) is inside a loop that *also* pushes items into the vector. So there could be several successful iterations, lengthening the vector, but final failure. At the point of failure there could be some number of no-longer-valid indexes that have been added to the vector, so clearing it is the only option (other than saving it before the loop and restoring it when the error condition is encountered...).

Reply:
And (to clarify a bit more), the reason that *that* is a problem is because at line 6759 below, we make a recursive call, but if it fails we don't return an error; instead *we continue executing in this function* (but now with a child_indexes vector that would be incorrect if the recursive call hadn't cleared it or restored it to its original state).

Reply:
So, my preferred solution would be to have this function return a `vector<int>` (apparently, there's even a TODO in the header about that (*)). That way you can build the result from the bottom up, and only add things to the vector when you've found the thing you're searching for and are certain that you're going to return successfully. That's going to be slightly slower than passing the same vector always (but maybe not slower than making a copy of it preemptively). However, I can live with both yours and Michael's solution (which I think could still work if you keep the `pop_back` statement so that the next iteration starts with a valid state). I'll leave it up to Michael to decide, as he spends most of the time looking at this code.

(*) Technically the TODO is about returning a `vector<vector<int>>` of all possible matches --  which may be a good idea overall, but I don't think we want to do that now.

Reply:
> So there could be several successful iterations, lengthening the vector, but final failure. At the point of failure there could be some number of no-longer-valid indexes that have been added to the vector, so clearing it is the only option (other than saving it before the loop and restoring it when the error condition is encountered...).
>
> And (to clarify a bit more), the reason that that is a problem is because at line 6759 below, we make a recursive call, but if it fails we don't return an error; instead we continue executing in this function (but now with a child_indexes vector that would be incorrect if the recursive call hadn't cleared it or restored it to its original state).

Right, but if we changed `GetIndexOfChildMemberWithName` to return an `Expected`, then we'd be able to handle the failure of the recursive call too. My thinking was that the case where Clang finds a path to a base-class but we then fail to compute the index to that element feels like a situation where we should bail out of this function. Although the counterpoint would be, what if Clang found paths to multiple base-classes, and only one of them failed. Then yes, we need the backtracking that you implement here. But looking current implementation, wouldn't we return a bogus result if Clang found multiple `CXXBasePaths` anyway? It would just add all the indices into `child_indexes` (it doesn't just pick "the first result" like the FIXME that Pavel points to suggests).

Correct me if I'm misunderstanding your points though!

I'm happy to go with your simple approach for now and clean up the function as a follow-up. It looks like there's a few things wrong with it. And the test-case you added here will be useful to drive that refactor.

I  think I prefer @labath's suggestion over mine too. That feels like the most robust option.